contrib/coredns/policy/policy.go

```diff
@@ -119,11 +120,39 @@ func (p *PolicyMiddleware) getEDNS0Attrs(r *dns.Msg) ([]*pb.Attribute, bool) {
 	return attrs, foundSourceIP
 }
 
+// This function validates the middleware response
+// Inputs is *PolicyMiddle, *Response and error
+// Output is DNS status and error
+func ValidateMiddlewareResponse(p *PolicyMiddleware, w dns.ResponseWriter, r *dns.Msg,response *Response, status int, err error) (int, error) {
```

Comment:
Validate is not the right name here, it's doing more than validating, it's actually modifying the request if needed. I think just put this into ServeDNS, it's only ever called once.